## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics

df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [6]:
# Use this space to do any additional research
# and familiarize yourself with the data.

print(df.info())

print(df.isna().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [7]:
# What three item categories had the most entries?

# What three item categories had the most entries?
most_entries_category = df['category'].value_counts()

most_entries_category.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [8]:
# For the category with the most entries,
# which subcategory had the most entries?
most_entries_subcategory = df.loc[df['category']=='consumables']
most_entries_subcategory['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [15]:
# Which five clients had the most entries in the data?
# Which five clients had the most entries in the data?
client_ids = df['client_id'].value_counts()
client_ids.head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [16]:
# Store the client ids of those top 5 clients in a list.

top_5_clientlist=client_ids.head(5).index.tolist()
top_5_clientlist

[33615, 66037, 46820, 38378, 24741]

In [17]:
# How many total units (the qty column) did the
# client with the most entries order order?

clientid=top_5_clientlist[0]
ordered_quantity = df.loc[df['client_id']==clientid]['qty'].sum()
print(ordered_quantity)

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [21]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['line_subtotal']=df['unit_price']*df['qty']



In [22]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

def calc_shippingprice(unit_weight):
    if unit_weight > 50.0: 
        return unit_weight*7
    else:
        return unit_weight*10


df['shipping_price']= (df['unit_weight']*df['qty']).apply(lambda row: calc_shippingprice(row))
df['total_weight'] = df['unit_weight']*df['qty']

In [23]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

df['total_price'] = (df['shipping_price']+df['line_subtotal'])*109.25/100

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  line_subtotal  shipping_price  total_weight  \
0         7.50  105            1      115164.00         5512.50        787.50   
1         1.49   21            0         523.95          312.90         31.29   
2         1.68   39            6         527.28          458.64         65.52   
3         1.23   29            3        1056.18          356.70         35.67   
4        46.43   20            1        3902.00         6500.20        928.60   

     total_price  
0  131839.076250  
1     914.258625  
2    1077.117600  
3    1543.571400  
4   11364.403500  

[5 rows x 22 columns]

In [24]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).


df['line_cost']= ( df['unit_cost']*df['qty'] )+ df['shipping_price'] 

df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  shipping_price  total_weight    total_price  \
0            1      115164.00         5512.50        787.50  131839.076250   
1            0         523.95          312.90         31.29     914.258625   
2            6         527.28          458.64         65.52    1077.117600   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [27]:
# Create a column for the profit of
# each line using line cost and line price

df['total_profit']=df['total_price']-df['line_cost']

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [29]:
# Check your work using the totals # Check your work using the totals above
df_order = df.loc[df['order_id']==2742071]
orderid=df_order['order_id'].unique()
totalcost=df_order['total_price'].sum().round(2)
print(f"Order ID # {orderid} , Total Price {totalcost}")

df_order = df.loc[df['order_id']==2173913]
orderid=df_order['order_id'].unique()
totalcost=df_order['total_price'].sum().round(2)
print(f"Order ID # {orderid} , Total Price {totalcost}")

df_order = df.loc[df['order_id']==6128929]
orderid=df_order['order_id'].unique()
totalcost=df_order['total_price'].sum().round(2)
print(f"Order ID # {orderid} , Total Price {totalcost}")


Order ID # [2742071] , Total Price 152811.9
Order ID # [2173913] , Total Price 162388.72
Order ID # [6128929] , Total Price 923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [30]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

top_5_clientlist
condition = df['client_id'].isin(top_5_clientlist)
df_spend = df.loc[condition]

grouped = df_spend.groupby('client_id')
grouped['total_price'].sum().round(2)

client_id
24741    82268891.98
33615     8377308.57
38378    12906550.88
46820     9743794.32
66037    10259514.80
Name: total_price, dtype: float64

In [32]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

sum_by_category = grouped[['qty', 'shipping_price','total_price','line_cost','total_profit']].sum().round(2)

sum_by_category

qty  shipping_price  total_price    line_cost  total_profit
client_id                                                                
24741      239862      5126448.37  82268891.98  45688899.71   36579992.27
33615       64313      1828984.89   8377308.57   6175313.91    2201994.66
38378       73667      3429455.40  12906550.88   9634720.98    3271829.90
46820       75768      1601448.84   9743794.32   7007191.64    2736602.68
66037       43018      1395151.85  10259514.80   7004482.98    3255031.82

In [39]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

sum_by_category.rename(index='Client ID')

sum_by_category.rename(columns={'qty':'Units','shipping_price':'Shipping','total_price':'Total Revenue','line_cost':'Total Cost','total_profit':'Total Profit'})

TypeError: Index(...) must be called with a collection of some kind, 'Client ID' was passed

In [19]:
# Sort the updated data by "Total Profit" form highest to lowest


Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M